flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.highavailability;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.HighAvailabilityOptions;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static org.apache.flink.util.StringUtils.isNullOrWhitespaceOnly;
+
+/**
+ * An implementation of the {@link JobResultStore} which persists job result data to an underlying
+ * distributed filesystem.
+ */
+public class FileSystemJobResultStore implements JobResultStore {
+
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    private final FileSystem fileSystem;
+
+    private final Path basePath;
+
+    private final boolean deleteOnCommit;
+
+    private static final String DIRTY_SUFFIX = "_DIRTY";
+
+    public FileSystemJobResultStore(Configuration config) throws IOException {
+        final String jrsStoragePath = config.get(JobResultStoreOptions.STORAGE_PATH);
+
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            this.basePath = new Path(haStoragePath + "/job-results-store/" + haClusterId);
+        } else {
+            this.basePath = new Path(jrsStoragePath);
+        }
+
+        this.fileSystem = this.basePath.getFileSystem();
+        this.fileSystem.mkdirs(this.basePath);
```

Comment:
It might be overkill, but: I was thinking about having wrapper around the FileSystem that takes care of creating and renaming files and checking for existence and dealing with the folder structure. `FileSystemJobResultStore` would be in charge of the serialization and utilizing this wrapper implementation. ...but that's just an idea I wanted to share. I haven't thought this through. ...especially considering that `FileSystemJobResultStore` is not too complex on its own. ¯\_(ツ)_/¯